File: Go/internal/api/router.go
package api

import (
	"github.com/gin-gonic/gin"
)

func SetupRouter(h *Handler) *gin.Engine {
	r := gin.Default()

	r.GET("/health", h.Health)

	// Users
	r.POST("/users", h.CreateUser)
	r.GET("/users", h.GetAllUsers)
	r.GET("/users/:id", h.GetUser)
	r.PUT("/users/:id", h.UpdateUser)
	r.DELETE("/users/:id", h.DeleteUser)

	// Tasks
	r.POST("/users/:id/tasks", h.CreateTask)
	r.GET("/users/:id/tasks", h.GetAllTasks)
	r.GET("/tasks/:tid", h.GetTask)
	r.PUT("/tasks/:tid", h.UpdateTask)
	r.PATCH("/tasks/:tid/done", h.MarkTaskDone)
	r.DELETE("/tasks/:tid", h.DeleteTask)

	return r
}

File: Bun/src/router.ts
import { Handlers } from "./handlers";

export class Router {
    private handlers: Handlers;

    constructor(handlers: Handlers) {
        this.handlers = handlers;
    }

    async handle(req: Request): Promise<Response> {
        const url = new URL(req.url);
        const path = url.pathname;
        const method = req.method;

        // Health
        if (path === "/health" && method === "GET") {
            return this.handlers.health(req);
        }

        // Users
        if (path === "/users" && method === "POST") {
            return this.handlers.createUser(req);
        }
        if (path === "/users" && method === "GET") {
            return this.handlers.getAllUsers(req);
        }

        // Match /users/:id
        const usersIdMatch = path.match(/^\/users\/(\d+)$/);
        if (usersIdMatch) {
            const id = parseInt(usersIdMatch[1]);
            if (method === "GET") return this.handlers.getUserById(req, id);
            if (method === "PUT") return this.handlers.updateUser(req, id);
            if (method === "DELETE") return this.handlers.deleteUser(req, id);
        }

        // Match /users/:id/tasks
        const userTasksMatch = path.match(/^\/users\/(\d+)\/tasks$/);
        if (userTasksMatch) {
            const userId = parseInt(userTasksMatch[1]);
            if (method === "POST") return this.handlers.createTaskForUser(req, userId);
            if (method === "GET") return this.handlers.getTasksByUserId(req, userId);
        }

        // Match /tasks/:id
        const tasksIdMatch = path.match(/^\/tasks\/(\d+)$/);
        if (tasksIdMatch) {
            const id = parseInt(tasksIdMatch[1]);
            if (method === "GET") return this.handlers.getTaskById(req, id);
            if (method === "PUT") return this.handlers.updateTask(req, id);
            if (method === "DELETE") return this.handlers.deleteTask(req, id);
        }

        // Match /tasks/:id/done
        const tasksDoneMatch = path.match(/^\/tasks\/(\d+)\/done$/);
        if (tasksDoneMatch) {
            const id = parseInt(tasksDoneMatch[1]);
            if (method === "PATCH") return this.handlers.markTaskDone(req, id);
        }

        return new Response("Not Found", { status: 404 });
    }
}
